### PR TITLE
feat(op-interop-filter): implement core filtering logic

### PR DESCRIPTION
**Pull request by opsuperchain**
# feat(op-interop-filter): implement core filtering logic

## Summary

This PR implements the core filtering logic for `op-interop-filter`, a lightweight interop validation service designed to run alongside sequencers. The filter validates cross-chain executing messages by:

- Ingesting blocks from multiple L2 chains into per-chain LogsDB databases
- Validating message timing constraints (init before execution, expiry windows)
- Tracking cross-validated timestamps across all chains using a lockstep model
- Providing a `supervisor_checkAccessList` RPC endpoint for sequencers to validate transactions

This implementation focuses on the filter logic and does not include sysgo/devstack integration, which is deferred to a follow-up PR.

## Architecture

```
                                 ┌─────────────────────────────────────────┐
                                 │                Service                  │
                                 │  (Lifecycle, config, pprof, metrics)    │
                                 └───────────────────┬─────────────────────┘
                                                     │
                    ┌────────────────────────────────┼────────────────────────────────┐
                    │                                │                                │
                    ▼                                ▼                                ▼
           ┌────────────────┐             ┌──────────────────┐             ┌──────────────────┐
           │  RPC Server    │             │     Backend      │             │  Metrics Server  │
           │  (geth RPC)    │             │  (coordinator)   │             │  (prometheus)    │
           └───────┬────────┘             └────────┬─────────┘             └──────────────────┘
                   │                               │
         ┌─────────┴─────────┐                     │
         │                   │                     │
         ▼                   ▼                     │
 ┌───────────────┐  ┌────────────────┐             │
 │ QueryFrontend │  │ AdminFrontend  │             │
 │ (supervisor)  │  │ (admin, JWT)   │             │
 └───────┬───────┘  └───────┬────────┘             │
         │                  │                      │
         └────────┬─────────┘                      │
                  │                                │
                  ▼                                ▼
           ┌─────────────────────────────────────────────────────────────────┐
           │                            Backend                              │
           │   - Coordinates chain ingesters and cross-validator             │
           │   - Manages failsafe mode (manual + error-based)                │
           │   - Routes CheckAccessList requests                             │
           └──────────────┬───────────────────────────────┬──────────────────┘
                          │                               │
                          ▼                               ▼
           ┌──────────────────────────────┐   ┌──────────────────────────────┐
           │   LockstepCrossValidator     │   │  ChainIngesters (per chain)  │
           │                              │   │                              │
           │  - Validates executing msgs  │◄──┤  - LogsDBChainIngester       │
           │  - Tracks cross-validated    │   │    (RPC + logsdb)            │
           │    timestamp across chains   │   │  - MockChainIngester         │
           │  - Lockstep advancement:     │   │    (in-memory for tests)     │
           │    waits for ALL chains      │   │                              │
           └──────────────────────────────┘   └──────────────┬───────────────┘
                                                             │
                                              ┌──────────────┴──────────────┐
                                              │                             │
                                              ▼                             ▼
                                       ┌────────────┐                ┌────────────┐
                                       │  Chain A   │                │  Chain B   │
                                       │  LogsDB    │                │  LogsDB    │
                                       │  (local)   │                │  (local)   │
                                       └────────────┘                └────────────┘
```

## Files Overview

| File | Purpose |
|------|---------|
| `interfaces.go` | Core interfaces: `ChainIngester` and `CrossValidator` |
| `backend.go` | Central coordinator managing chain ingesters and failsafe state |
| `logsdb_chain_ingester.go` | Production `ChainIngester`: RPC + logsdb + reorg detection |
| `lockstep_cross_validator.go` | Production `CrossValidator`: timestamp advancement + validation |
| `frontend.go` | RPC frontends: public query + JWT-protected admin |
| `service.go` | Top-level lifecycle orchestration |
| `config.go` | CLI flag parsing |
| `validation.go` | Message timing validation |
| `errors.go` | Ingester error types |
| `mock_ingester_test.go` | In-memory `ChainIngester` for unit tests |
| `backend_test.go` | Comprehensive unit tests |

## Review Guide

**Recommended order:**

1. **`interfaces.go`** - Core abstractions (`ChainIngester`, `CrossValidator`)
2. **`backend.go`** - How components connect
3. **`lockstep_cross_validator.go`** - Cross-chain validation logic
4. **`logsdb_chain_ingester.go`** - Block ingestion and reorg detection
5. **`frontend.go`** - RPC layer
6. **`backend_test.go`** - Validate understanding through tests

## Test Coverage

- **Message Timing**: init < inclusion, expiry boundaries
- **Failsafe**: manual toggle, automatic on chain errors
- **Ready State**: chain readiness tracking
- **CheckAccessList**: failsafe, uninitialized, safety levels
- **CrossValidator**: timeout expiry, cross-unsafe timestamps, unknown chains, checksum verification, error propagation

## Future Improvements

1. **Reorg handling** - Recovery logic for chain reorganizations
2. **Specific invalid message flagging** - See TODO in `lockstep_cross_validator.go:217`
3. **Recoverable error states** - Recover without full restart
4. **Kurtosis/devstack integration tests** - Follow-up PR
5. **Per-chain validation advancement** - Remove lockstep bottleneck

## Test Plan

- [x] Unit tests pass
- [x] Manual testing with `interop-filter-utils`
- [ ] Kurtosis/devstack integration (follow-up PR)

---

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by karlfloersch** [comment]
@codex pls review

File: op-interop-filter/filter/backend.go [hunk 1]
```diff
@@ -2,55 +2,472 @@ package filter
...
+	// Create chain ingesters for each L2 RPC
+	for _, rpcURL := range cfg.L2RPCs {
+		// Query chain ID from the RPC
+		chainID, err := b.queryChainID(ctx, rpcURL)
+		if err != nil {
+			cleanup()
+			return nil, fmt.Errorf("failed to query chain ID from %s: %w", rpcURL, err)
+		}
+
+		// Check for duplicate chain IDs BEFORE creating ingester
```
Comment (karlfloersch) [inline op-interop-filter/filter/backend.go]: No need to all caps BEFORE

**Comment by karlfloersch** [comment]
Add a note for how to improve the cross validation logic to be faster (not wait for global cross unsafe to advance, instead do on a per chain basis)

**Comment by opsuperchain** [comment]
# Review Guide: PR #18604 Refactoring Commit

This commit addresses all review feedback from the upstream PR. Below is a comprehensive guide for reviewing the changes.

## Quick Stats
- **Files changed:** 15
- **Lines:** +2,268 / -675
- **Test functions:** 43 total (38 top-level + 5 subtests)

---

## High-Level Changes

### 1. Validation Logic Consolidation
**Files:** `lockstep_cross_validator.go`, `validation.go` (deleted)

Moved all validation logic into the cross-validator. This eliminates the separate `validation.go` file and makes the validation path clearer.

### 2. Ingestion Loop Improvements  
**Files:** `logsdb_chain_ingester.go`

- Replaced `time.Sleep` polling with `time.Ticker` for cleaner shutdown
- Added parent hash validation on every block (catches reorgs immediately)
- Added constants for magic numbers (`progressLogInterval`, `dataDirPermissions`)
- Improved `checkReorg()` to distinguish expected vs unexpected cases:
  - Head before earliest block → expected, debug log (we started later)
  - Have block but can't get hash → unexpected, error log
  - Hash mismatch → reorg detected, trigger failsafe

### 3. Safety: Panic → Error Returns
**Files:** `lockstep_cross_validator.go:137-143, 193-196`

Replaced 3 panics with proper error returns to prevent DoS via crafted timestamps. Example:
```go
// Before (panic reachable from RPC)
if expiresAt < access.Timestamp {
    panic("overflow in expiry calculation")
}

// After (returns error)
if expiresAt < access.Timestamp {
    return fmt.Errorf("overflow in expiry calculation: %w", types.ErrConflict)
}
```

### 4. Dead Code Removal
**Files:** `logsdb_chain_ingester.go`

Deleted unused `findBlockByTimestamp()` binary search function (50 lines).

### 5. Error Handling Improvements
**Files:** `errors.go`, `logsdb_chain_ingester.go`

- Added `ErrInvalidLog` sentinel error
- Added `ErrorInvalidExecutingMessage` ingester error reason
- Added failsafe triggers for `ErrDataCorruption` and `ErrInvalidExecutingMessage`

### 6. Config Integration
**Files:** `config.go`, `flags/flags.go`, `service.go`

Added `--networks` and `--rollup-configs` flags for chain configuration via superchain-registry.

---

## Test Coverage: Complete Breakdown

### Backend Tests (`backend_test.go`) - 6 tests

| Test | Code Path | What It Verifies |
|------|-----------|------------------|
| `TestBackend_Failsafe_ManualEnabled` | `SetFailsafeEnabled()`, `FailsafeEnabled()` | Manual failsafe toggle works |
| `TestBackend_Failsafe_ChainError` | `FailsafeEnabled()` with chain error | Chain ingester errors trigger failsafe |
| `TestBackend_Failsafe_CrossValidatorError` | `FailsafeEnabled()` with validator error | Validator errors trigger failsafe |
| `TestBackend_Failsafe_AllClear` | `FailsafeEnabled()` normal state | No errors = no failsafe |
| `TestBackend_Ready` | `Ready()` | Backend ready state with/without chains |
| `TestBackend_CheckAccessList` | `CheckAccessList()` | Failsafe, not ready, safety levels, unknown chain, happy path |

### Cross-Validator Tests (`lockstep_cross_validator_test.go`) - 12 tests

| Test | Code Path | What It Verifies |
|------|-----------|------------------|
| `TestCrossValidator_TimeoutExceedsExpiry` | Expiry calculation | Rejects if timeout > expiry window |
| `TestCrossValidator_CrossUnsafe_Boundary` | CrossUnsafe validation | At boundary accepts, beyond rejects |
| `TestCrossValidator_KnownChain` | Chain lookup | Succeeds for known chains |
| `TestCrossValidator_UnknownChain` | Chain lookup | Returns error for unknown chains |
| `TestCrossValidator_ValidationFailureSetsError` | Error state | Validation failure sets validator error |
| `TestValidateAccessEntry_TimestampNotIngested` | Min ingested check | Rejects if timestamp not yet ingested |
| `TestValidateExecMsg_InitBeforeInclusion` | Temporal validation | Rejects when init timestamp not before inclusion |
| `TestValidateExecMsg_MessageExpired` | Expiry validation | Rejects expired messages |
| `TestAdvanceValidation_WaitsForChainsReady` | `advanceValidation()` | Waits until all chains ready |
| `TestAdvanceValidation_InitializesToStartTimestamp` | `advanceValidation()` | Initializes to start timestamp |
| `TestAdvanceValidation_AdvancesTimestamp` | `advanceValidation()` | Advances crossValidatedTs over time |
| `TestAdvanceValidation_StopsAtMinIngested` | `advanceValidation()` | Stops at slowest chain timestamp |

### Chain Ingester Tests (`logsdb_chain_ingester_test.go`) - 18 tests

| Test | Code Path | What It Verifies |
|------|-----------|------------------|
| `TestLogsDBChainIngester_InitLogsDB` | `initLogsDB()` | Creates logsDB and chain directory |
| `TestLogsDBChainIngester_SealParentBlock` | `sealParentBlock()` | Seals parent block from RPC |
| `TestLogsDBChainIngester_IngestBlock` | `ingestBlock()` | Ingests block and stores logs |
| `TestLogsDBChainIngester_IngestMultipleBlocks` | `ingestBlock()` x N | Sequential block ingestion |
| `TestLogsDBChainIngester_Ready` | `Ready()` | Reports ready after reaching start timestamp |
| `TestLogsDBChainIngester_ErrorState` | `SetError()`, `Error()`, `ClearError()` | Error state management works |
| `TestLogsDBChainIngester_Contains` | `Contains()` | Queries stored logs correctly |
| `TestLogsDBChainIngester_ReorgDetection` | Parent hash validation | Detects reorg via parent hash mismatch |
| `TestLogsDBChainIngester_QueryMethods` | `GetExecMsgsAtTimestamp()`, `BlockHashAt()` | Query methods work correctly |
| `TestLogsDBChainIngester_Integration_RealLogsDB` | Full flow | End-to-end ingestion with real logsDB |
| `TestLogsDBChainIngester_InitIngestion_FreshStart` | `initIngestion()` fresh | Starts from calculated backfill block |
| `TestLogsDBChainIngester_InitIngestion_ResumeFromExistingDB` | `initIngestion()` resume | Resumes from existing DB state |
| `TestLogsDBChainIngester_InitIngestion_ErrorGettingHead` | `initIngestion()` RPC error | Handles RPC failure gracefully |
| `TestLogsDBChainIngester_IngestBlock_NonSequentialError` | Block ordering | Rejects non-sequential blocks |
| `TestLogsDBChainIngester_IngestBlock_RPCError` | RPC failure | Handles missing block gracefully |
| `TestLogsDBChainIngester_IngestBlock_ReceiptsError` | Receipts failure | Handles missing receipts gracefully |
| `TestLogsDBChainIngester_IngestBlock_ErrorStateSkipsIngestion` | Error state | Skips ingestion when in error state |
| `TestLogsDBChainIngester_ErrorTypes` | Error type strings | New error reasons have correct strings |

### JWT Auth Tests (`jwt_auth_test.go`) - 2 tests (5 subtests)

| Test | Code Path | What It Verifies |
|------|-----------|------------------|
| `TestJWTAuthentication` | JWT middleware | Valid JWT allows access, invalid rejected, supervisor API not gated |
| `TestSupervisorAPIWithoutJWT` | JWT middleware | Supervisor API works without JWT configured |

---

## Testing Approach

### Mock Strategy

**`mockChainIngester`** - In-memory implementation for unit tests:
- Stores logs in a `map[logKey]BlockSeal`
- Simple state management (`ready`, `err`, `latestBlock`)
- Thread-safe with `sync.RWMutex`
- Used by: Backend tests, Cross-validator tests

**`MockEthClient`** - RPC mock for integration tests:
- Stores blocks and receipts in maps
- Supports error injection (`SetInfoByLabelErr`, etc.)
- Used by: LogsDBChainIngester tests

**`mockCrossValidator`** - Simple validator mock:
- Returns configurable errors
- Used by: Backend tests

### Test Categories

1. **Unit tests** - Test individual methods with mocks
2. **Integration tests** - Test component interactions with real logsDB
3. **Error injection tests** - Verify graceful handling of failures
4. **Boundary tests** - Test edge cases (overflow, expiry, timestamps)

### What's NOT Tested (Known Gaps)

- `Start()`/`Stop()` lifecycle (requires goroutine coordination)
- `runIngestion()` loop (tested indirectly via `ingestBlock()`)
- Concurrent access (no explicit race tests, but uses atomic/mutex)

---

## Review Checklist

- [ ] Validation logic correctly moved from `validation.go` to cross-validator
- [ ] Panics replaced with error returns (3 locations)
- [ ] Dead code removed (`findBlockByTimestamp`)
- [ ] New error types used correctly (`ErrDataCorruption`, `ErrorInvalidExecutingMessage`)
- [ ] Tests cover the documented code paths
- [ ] Mock naming is consistent (`mockChainIngester`, `mockCrossValidator`, `MockEthClient`)

File: op-interop-filter/filter/logsdb_chain_ingester.go [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+			return nil
+		}
+		if errors.Is(err, types.ErrDataCorruption) {
+			c.SetError(ErrorDataCorruption, fmt.Sprintf("data corruption at block %d: %v", blockNum, err))
+			return nil
+		}
+		if errors.Is(err, ErrInvalidLog) {
+			c.SetError(ErrorInvalidExecutingMessage, fmt.Sprintf("invalid log at block %d: %v", blockNum, err))
+			return nil
+		}
```
Comment (karlfloersch) [inline op-interop-filter/filter/logsdb_chain_ingester.go:633]: Remove

File: op-interop-filter/filter/logsdb_chain_ingester.go [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+	}
+
+	for _, receipt := range receipts {
+		for _, l := range receipt.Logs {
+			logHash := processors.LogToLogHash(l)
+
+			execMsg, err := processors.DecodeExecutingMessageLog(l)
+			if err != nil {
+				return 0, fmt.Errorf("invalid log %d in block %d: %w: %w", l.Index, blockNum, ErrInvalidLog, err)
+			}
```
Comment (karlfloersch) [inline op-interop-filter/filter/logsdb_chain_ingester.go:672]: Remove all logHash or execMsg checks

**Comment by opsuperchain** [comment]
Closing: base branch targets upstream develop
